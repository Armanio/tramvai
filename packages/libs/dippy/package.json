{
  "name": "@tinkoff/dippy",
  "version": "0.0.0-stub",
  "initialVersion": "0.7.27",
  "description": "",
  "main": "./lib/di.js",
  "typings": "./lib/di.d.ts",
  "files": [
    "__migrations__",
    "lib"
  ],
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git@github.com:TinkoffCreditSystems/tramvai.git"
  },
  "scripts": {
    "build": "tramvai-build",
    "watch": "tsc -w",
    "build-for-publish": "true"
  },
  "gitHead": "8e826a214c87b188fc4d254cdd8f2a2b2c55f3a8",
  "dependencies": {
    "tslib": "^2.0.3"
  },
  "module": "./lib/di.es.js",
  "es2017": "./lib/di.es2017.js",
  "license": "Apache-2.0"
}
