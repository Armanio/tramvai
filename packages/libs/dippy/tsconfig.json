{
  "extends": "../../../tsconfig.lib.json",
  "include": [
    "./src"
  ],
  "compilerOptions": {
    "module": "CommonJS",
    "declarationDir": "./lib",
    "outDir": "./lib",
    "rootDir": "./src",
    "strict": false,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "strictBindCallApply": false,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true
  }
}
