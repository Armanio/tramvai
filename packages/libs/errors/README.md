# Errors

Основные классы ошибок для проектов Tinkoff

## RedirectFoundError

Обозначает что был обнаружен редирект в рамках приложения

## NotFoundError

Обозначает что страница не была найдена в приложении
