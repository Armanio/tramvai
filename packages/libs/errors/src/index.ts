import { NotFoundError } from './notFound';
import { RedirectFoundError } from './redirect';
import { HttpError } from './httpError';

export const throwNotFoundError = (parameters?: ConstructorParameters<typeof NotFoundError>[0]) => {
  throw new NotFoundError(parameters);
};

export const throwRedirectFoundError = (
  parameters: ConstructorParameters<typeof RedirectFoundError>[0]
) => {
  throw new RedirectFoundError(parameters);
};

export const throwHttpError = (parameters: ConstructorParameters<typeof HttpError>[0]) => {
  throw new HttpError(parameters);
};

export const isNotFoundError = (err: Error): err is NotFoundError => {
  return err.name === NotFoundError.errorName;
};

export const isRedirectFoundError = (err: Error): err is RedirectFoundError => {
  return err.name === RedirectFoundError.errorName;
};

export const isHttpError = (err: Error): err is HttpError => {
  return err.name === HttpError.errorName || isNotFoundError(err) || isRedirectFoundError(err);
};

export { NotFoundError, RedirectFoundError, HttpError };
