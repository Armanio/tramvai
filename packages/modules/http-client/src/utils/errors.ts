export const getErrorMessage = (message: string) => {
  switch (true) {
    case /network timeout at/i.test(message) || /request timed out/i.test(message):
      return 'Время истекло. Обновите страницу и повторите попытку.';
    default:
      return 'Что-то пошло не так. Мы уже решаем проблему. Попробуйте снова через несколько минут.';
  }
};

export const formatError = (error: Error & { originalMessage?: string }) => {
  if (error.originalMessage) {
    return error;
  }

  const originalMessage = error.message;
  const message = getErrorMessage(originalMessage);

  return Object.assign(error, { message, originalMessage });
};
