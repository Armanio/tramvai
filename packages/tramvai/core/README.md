# Core

Ядро tramvai. В основном требуется для разработки модулей трамвая.

## Экспортируемые токены

### DI_TOKEN

Используется для доступа к инстансу DI

### APP_INFO_TOKEN

Содержит информацию о запущенном приложении

### COMMAND_LINE_RUNNER_TOKEN

Реализация раннера для исполнения команд

### COMMAND_LINES_TOKEN

Задание списка команд
