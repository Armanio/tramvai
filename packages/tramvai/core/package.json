{
  "name": "@tramvai/core",
  "version": "0.0.0-stub",
  "description": "",
  "main": "lib/index.js",
  "typings": "src/index.ts",
  "files": [
    "lib",
    "__migrations__"
  ],
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git@github.com:TinkoffCreditSystems/tramvai.git"
  },
  "scripts": {
    "build": "tramvai-build",
    "watch": "tsc -w",
    "build-for-publish": "true"
  },
  "dependencies": {
    "@tinkoff/utils": "^2.1.2",
    "@tramvai/types-actions-state-context": "0.0.0-stub",
    "@tramvai/tools-check-versions": "0.0.0-stub",
    "@tramvai/tools-migrate": "0.0.0-stub",
    "@tinkoff/dippy": "0.0.0-stub",
    "tslib": "^2.0.3"
  },
  "peerDependencies": {
    "jscodeshift": "^0.7.0"
  },
  "gitHead": "8e826a214c87b188fc4d254cdd8f2a2b2c55f3a8",
  "module": "lib/index.es.js",
  "es2017": "lib/index.es2017.js",
  "license": "Apache-2.0"
}
