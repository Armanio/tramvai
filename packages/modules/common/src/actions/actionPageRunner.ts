import type { Action } from '@tramvai/core';
import { ACTION_PARAMETERS } from '@tramvai/core';
import type {
  LOGGER_TOKEN,
  ActionPageRunner as ActionPageRunnerInterface,
  STORE_TOKEN,
} from '@tramvai/tokens-common';
import { actionServerStateEvent } from './actionTramvaiReducer';
import type { ActionExecution } from './actionExecution';
import { actionType } from './constants';

const payload = {};

export class ActionPageRunner implements ActionPageRunnerInterface {
  store: typeof STORE_TOKEN;

  actionExecution: ActionExecution;

  private limitTime: number;

  private log: ReturnType<typeof LOGGER_TOKEN>;

  constructor({ store, actionExecution, limitTime, logger }) {
    this.store = store;
    this.actionExecution = actionExecution;
    this.limitTime = limitTime;
    this.log = logger('action:action-page-runner');
  }

  // TODO stopRunAtError нужен только для редиректов на стороне сервера в экшенах. И нужно пересмотреть реализацию редиректов
  runActions(actions: Action[], stopRunAtError: (error: Error) => boolean = () => false) {
    return new Promise<void>((resolve, reject) => {
      const timeoutMarker = setTimeout(() => {
        this.log.warn(
          `page actions has exceeded timeout of ${this.limitTime}ms, ignore some results of execution`
        );
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        endChecker();
      }, this.limitTime);

      const endChecker = () => {
        clearTimeout(timeoutMarker);
        const result: Record<string, any> = {};
        this.actionExecution.execution.forEach((value, key) => {
          // достаем значения экшенов, которые успешно выполнились, остальные выполнятся на клиенте
          if (value.status === 'success') {
            result[key] = value;
          }
        });

        this.syncStateActions(result);
        resolve();
      };

      const actionMapper = (action: Action) => {
        return Promise.resolve()
          .then(() => this.actionExecution.run(action, payload, actionType.global))
          .catch((error) => {
            const parameters = action[ACTION_PARAMETERS];

            this.log.error({
              error,
              event: `action-execution-error`,
              message: `${parameters.name} execution error`,
            });

            if (stopRunAtError(error)) {
              clearTimeout(timeoutMarker);
              reject(error);
            }
          });
      };

      // eslint-disable-next-line promise/catch-or-return
      Promise.all(actions.map(actionMapper)).then(endChecker);
    });
  }

  private syncStateActions(success: Record<string, any>) {
    return this.store.dispatch(actionServerStateEvent(success));
  }
}
