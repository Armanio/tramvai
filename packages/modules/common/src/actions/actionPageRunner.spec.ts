import noop from '@tinkoff/utils/function/noop';
import { createAction } from '@tramvai/core';
import { throwRedirectFoundError } from '@tinkoff/errors';
import { createContainer } from '@tinkoff/dippy';
import { ActionExecution } from './actionExecution';
import { ActionPageRunner } from './actionPageRunner';

const contextMock = {};

const logger = () => ({
  warn: noop,
  error: noop,
});

const delay = (time) => new Promise((resolve) => setTimeout(resolve, time));

describe('actionPageRunner', () => {
  it('Запуск множество page действий без каких либо опций', async () => {
    const store: any = { getState: () => ({}), dispatch: () => {} };
    const instanceExecution = new ActionExecution({
      di: createContainer(),
      store,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
    });
    const instance = new ActionPageRunner({
      logger,
      actionExecution: instanceExecution,
      store,
      limitTime: 500,
    });
    const result = [];
    const action1 = createAction({
      fn: (c, payload) => result.push(['action1', payload]),
      name: 'action1',
    });
    const action2 = createAction({
      fn: async (c, payload) => {
        result.push(['action2', payload]);

        await instanceExecution.run(action1, 1);
        return instanceExecution.run(action1, 2);
      },
      name: 'action2',
    });
    await instance.runActions([action1, action2]);

    expect(result).toEqual([
      ['action1', {}],
      ['action2', {}],
      ['action1', 1],
      ['action1', 2],
    ]);
  });

  // TODO Тест работает не верно
  it('Выполнение действий заняло больше установленных значений', async () => {
    let dispatcherPayload;
    const store: any = {
      getState: () => ({}),
      dispatch: (data) => {
        dispatcherPayload = data;
      },
    };
    const instanceExecution = new ActionExecution({
      di: createContainer(),
      store,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
    });
    const instance = new ActionPageRunner({
      logger,
      actionExecution: instanceExecution,
      store,
      limitTime: 500,
    });
    const result = [];
    const prom = instance.runActions([
      createAction({
        fn: () => delay(2000).then(() => result.push(1)),
        name: 'action time 1',
      }),
      createAction({
        fn: () => delay(2000).then(() => result.push(2)),
        name: 'action time 2',
      }),
      createAction({
        fn: () => result.push(3),
        name: 'action time 3',
      }),
    ]);

    // ждем когда отрезовлится action time 3
    Promise.resolve().then(() => {
      jest.runOnlyPendingTimers();
    });

    return prom.then(() => {
      expect(result).toEqual([3]);
      expect(dispatcherPayload).toMatchSnapshot();
    });
  });

  it('Упавшие экшены должны быть исключены из стора', async () => {
    let dispatcherPayload;
    const store: any = {
      getState: () => ({}),
      dispatch: (data) => {
        dispatcherPayload = data;
      },
    };
    const instanceExecution = new ActionExecution({
      di: createContainer(),
      store,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
    });
    const instance = new ActionPageRunner({
      logger,
      actionExecution: instanceExecution,
      store,
      limitTime: 500,
    });
    const result = [];
    await instance.runActions([
      createAction({ fn: () => Promise.reject(new Error('Error')), name: 'action1' }),
      createAction({ fn: () => result.push(2), name: 'action2' }),
    ]);

    expect(result).toEqual([2]);
    expect(dispatcherPayload).toMatchSnapshot();
  });

  describe('Перехват ошибок', () => {
    it('runner не должен падать при ошибках', async () => {
      const store: any = { getState: () => ({}), dispatch: () => {} };
      const instanceExecution = new ActionExecution({
        di: createContainer(),
        store,
        actionConditionals: [],
        // @ts-ignore
        context: contextMock,
      });
      const instance = new ActionPageRunner({
        logger,
        actionExecution: instanceExecution,
        store,
        limitTime: 500,
      });

      expect.assertions(1);

      return instance
        .runActions([
          createAction({
            fn: () =>
              Promise.resolve(1).then(() => {
                throw new Error('213');
              }),
            name: 'action1',
          }),
        ])
        .then((data) => {
          expect(true).toBe(true);
        });
    });

    it('При кастомных обработка ошибка раннер падает', async () => {
      const store: any = { getState: () => ({}), dispatch: () => {} };
      const instanceExecution = new ActionExecution({
        di: createContainer(),
        store,
        actionConditionals: [],
        // @ts-ignore
        context: contextMock,
      });
      const instance = new ActionPageRunner({
        logger,
        actionExecution: instanceExecution,
        store,
        limitTime: 500,
      });

      const stopRunAtError = (error) => {
        if (error.name === 'RedirectFoundError') {
          return true;
        }
      };

      expect.assertions(1);

      return instance
        .runActions(
          [
            createAction({
              fn: () => Promise.resolve(1).then(() => throwRedirectFoundError({ nextUrl: '' })),
              name: 'action1',
            }),
          ],
          stopRunAtError
        )
        .catch((error) => {
          expect(error.name).toBe('RedirectFoundError');
        });
    });
  });
});
