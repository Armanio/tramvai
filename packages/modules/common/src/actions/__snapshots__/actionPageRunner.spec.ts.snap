// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionPageRunner Выполнение действий заняло больше установленных значений 1`] = `
Object {
  "payload": Object {},
  "type": "action state execution in server",
}
`;

exports[`actionPageRunner Упавшие экшены должны быть исключены из стора 1`] = `
Object {
  "payload": Object {
    "action2": Object {
      "state": Object {},
      "status": "success",
    },
  },
  "type": "action state execution in server",
}
`;
